### ISSUE TITLE
Limit `git autofixup` to specified files and/or to staged changes

### ISSUE THREAD

**Issue by intelfx** [torbiak/git-autofixup#30]
It would be nice if it was possible to limit `git-autofixup` to changes in specific files, e. g.:

```
git autofixup <base> -- /path/to/file /path/to/something/else
```

Alternatively (might be even more useful), accept something like

- `--staged` to only operate on _already staged_ changes (mirroring `git stash --staged`), and
- `--keep-index` for the opposite (mirroring `git stash --keep-index`).

**Comment by krobelus** [comment]
On Sun, Feb 25, 2024 at 04:27:08AM -0800, Ivan Shapovalov wrote:
> It would be nice if it was possible to limit `git-autofixup` to changes in specific files, e. g.:
> 
> ```
> git autofixup <base> -- /path/to/file /path/to/something/else
> ```

Good idea. It is probably doable to support arbitrary pathspecs by forwarding them to Git diff.

> 
> Alternatively (might be even more useful), accept something like
> 
> - `--staged` to only operate on _already staged_ changes (mirroring `git stash --staged`), and

that's already the default in the latest release - if you have anything
staged, stage everything. It is indeed very useful.

> - `--keep-index` for the opposite (mirroring `git stash --keep-index`).

interesting, yeah. Would probably need to temporarily swap the index file but we already do that elsewhere.

**Comment by torbiak** [comment]
Regarding `--keep-index`: we could get conflicts when trying to restore the original index after doing git-autofixup for the working tree, right?

If I'm understanding correctly, it'd be roughly equivalent to this:

```
git checkout -b tmp
git commit -m 'saved index'
git checkout topic
git-autofixup
git checkout tmp
git rebase topic
git reset @^
```

But what if rebasing the saved index onto the changed topic branch fails?

**Comment by torbiak** [comment]
oh, I guess it'd be a lot nicer to write a `--keep-index` script as:
```
git stash push --staged
git-autofixup
git stash pop
```
Thanks for the suggestions, @intelfx, but to me it seems more flexible, tractable, and lower maintenance for people to write this as a shell script instead of working it into git-autofixup. Apart from convenience, is there an additional advantage that would be gained by having it in git-autofixup?

**Comment by torbiak** [comment]
@fasterit, thanks for the PR. It looks great except that it's ambiguous whether the first arg is a revision or file, and it doesn't seem too unusual to have a branch named the same as a file, like if you have both a dir and branch named after a project component. I've been thinking of ways to remove the ambiguity without breaking the current interface. The best I've come up with is to enable listing files if the revision has been explicitly given via an `-r` flag, but I'm having trouble coming up with a good sentinel value to indicate the default revision should be chosen. It'd be nice if it was short and couldn't collide with branch names, but a bunch of symbols are already used by git to mean various things. Like, `-` might be a good option, but in the context of specifying revisions it already has connotations of  "previous branch" from `git switch` and `git checkout`. Maybe `+`? What do you think?

**Comment by fasterit** [comment]
I had the same though process as you. Just I concluded that having revision names that match files is very unlikely. And keeping people's muscle memory is most important, so no `-r revision` or the like.

Originally I had implemented

```bash
git-autofixup [opts] revision -- file1 file2
```

with `--` being the common (optional) delimiter to allow filenames like `--inamemyfilesstupidly`. GetOptions supports this but it need a hack around the fact that `revision` is a non-option argument before the `--`.

I have resurrected the code and pushed it to the PR branch. May be you like it that way better?

**Comment by torbiak** [comment]
Yeah, I prefer your earlier approach. There's a bug with counting args, though: `$args_before_separator++ unless $arg =~ /^-/;` counts option values, too, like the 3 in `--strict 3`. I think we can avoid needing to detect options and option values ourselves by instead removing everything after `--` from `@ARGV` before calling `GetOptions()`, like in the following. I think this behaves comparably to git like for `git checkout (-p|--patch) [<tree-ish>] [--] [<pathspec>...]`. I looked at using Getopt::Long's "argument callback" but we'd need to enable `permute` and `pass_through` and handle bad options ourselves, then.

`opts.pl`:

```
#!/usr/bin/perl -w
use strict;
use Getopt::Long;
use Data::Dumper;

$Data::Dumper::Terse = 1;

my $strict;
my $raw_upstream;
my @files = ();

my $found_option_terminator = 0;
my @new_argv = ();
for (@ARGV) {
    if ($found_option_terminator) {
        push @files, $_;
    } elsif ($_ eq '--') {
        $found_option_terminator = 1;
    } else {
        push @new_argv, $_;
    }
}
@ARGV = @new_argv;

GetOptions(
    'strict|s=i' => \$strict,
);

if (@ARGV) {
    $raw_upstream = shift @ARGV;
    @files = (@ARGV, @files);
}

print Dumper({
    'strict' => $strict,
    'raw_upstream' => $raw_upstream,
    'files' => \@files,
});
```

And so:
```
$ ./opts.pl -s 3 -- a b
{
  'strict' => 3,
  'files' => [
               'a',
               'b'
             ],
  'raw_upstream' => undef
}

$ ./opts.pl -s 3 rev -- a b
{
  'files' => [
               'a',
               'b'
             ],
  'raw_upstream' => 'rev',
  'strict' => 3
}

$ ./opts.pl -s 3 rev a b
{
  'raw_upstream' => 'rev',
  'files' => [
               'a',
               'b'
             ],
  'strict' => 3
}
```

Are you interested in fixing that bug and adding a test where an option is given a value?

**Comment by fasterit** [comment]
Yes, I like your logic. That seems to be more easy and correct.
I've implemented tests first and then adjusted the option parsing logic as per your code above.
Pls. check.
